File: lib/models/fortnight.dart
class Fortnight {
  final int index;
  final DateTime startDate;
  final DateTime endDate;

  Fortnight({
    required this.index,
    required this.startDate,
    required this.endDate,
  });

  factory Fortnight.fromIndex(int index) {
    final DateTime baseDate = DateTime(2024, 0, 1);
    final startDate = baseDate.add(Duration(days: (index - 1) * 14));
    final endDate = startDate.add(const Duration(days: 13));
    return Fortnight(index: index, startDate: startDate, endDate: endDate);
  }

  factory Fortnight.fromDate(DateTime givenDate) {
    final DateTime baseStartDate =
        DateTime(2024, 0, 1); // Start date of FN1 is January 1, 2024

    // Calculate the difference in days
    const millisecondsPerDay = 1000 * 60 * 60 * 24;
    final deltaDays = ((givenDate.millisecondsSinceEpoch -
                baseStartDate.millisecondsSinceEpoch) /
            millisecondsPerDay)
        .floor();

    // Calculate FN period
    final fnPeriod = (deltaDays / 14).floor() + 1;
    return Fortnight.fromIndex(fnPeriod);
  }

  static List<Fortnight> fromRange(DateTime start, DateTime end) {
    final fortnights = <Fortnight>[];
    var current = DateTime(start.year, start.month, start.day);
    while (current.isBefore(end) || current.isAtSameMomentAs(end)) {
      final fortnight = Fortnight.fromDate(current);
      if (!fortnights.contains(fortnight)) {
        fortnights.add(fortnight);
      }
      current = current.add(const Duration(days: 1));
    }
    return fortnights;
  }

  factory Fortnight.fromJson(Map<String, dynamic> json) {
    return Fortnight.fromIndex(json['index']);
  }

  Map<String, dynamic> toJson() {
    return {
      'index': index,
      'startDate': startDate,
      'endDate': endDate,
    };
  }
}

File: functions/src/models/fortnight.ts
export class Fortnight {
    index: number;
    startDate: Date;
    endDate: Date;

    constructor(index: number) {
        this.index = index;
        const baseDate = new Date(2024, 0, 1);
        this.startDate = new Date(baseDate.setDate(baseDate.getDate() + (index - 1) * 14));
        this.endDate = new Date(baseDate.setDate(baseDate.getDate() + 13));


    }

    static fromDate(givenDate: Date): Fortnight {
        const baseStartDate: Date = new Date(2024, 0, 1); // Start date of FN1 is January 1, 2024

        // Calculate the difference in days
        const millisecondsPerDay = 1000 * 60 * 60 * 24;
        const deltaDays = Math.floor((givenDate.getTime() - baseStartDate.getTime()) / millisecondsPerDay);

        // Calculate FN period
        const fnPeriod = Math.floor(deltaDays / 14) + 1;
        return new Fortnight(fnPeriod);
    }

    static fromRange(start: Date, end: Date): Fortnight[] {
        const fortnights: Fortnight[] = [];
        const current = new Date(start);
        while (current <= end) {
            const fortnight = Fortnight.fromDate(current);
            const alreadyAdded = fortnights.some((fn) => fn.index === fortnight.index);
            if (!alreadyAdded) {
                fortnights.push(fortnight);
            }
            current.setDate(current.getDate() + 1);
        }
        return fortnights;
    }
}